### PR TITLE
Preparation for code review, ruby enumerable methods

### PR DESCRIPTION
**Pull request by jamezjaz**
This project is the ruby enumerable methods; on this branch, I created my own version of the methods which behaves just like the already existing enumerable methods. I created the following methods:
#my_each
#my_each_with_index
#my_select
#my_all?
#my_any?
#my_none?
#my_count
#my_map, #my_map_with_proc, #my_map_with_proc_or_block
#my_inject and #multiply_els which multiplies all the elements of the array together by using #my_inject

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress :+1: 
Nice work with the enumerables project. Let's fix some failing edge cases before we can merge and proceed.
- read my comments on suggestions for fixing the failing edge cases.
- enable stickler and make sure all test cases pass

### Martin Afani

File: my_enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    new_arr
+  end
+
+  def my_all?
+    result = true
+    for item in self do
+      return false if yield(item) == false
+    end
+    result
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
```ruby
true_array = [1, true, 'hi', []] 
false_array = [1, false, 'hi', []] 
true_array.my_all? == true_array.all? #true
```
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100){rand(0...9)}
array.my_all?(Integer) == array.all?(Integer)
```
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true
```

File: my_enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    end
+    false
+  end
+
+  def my_none?
+    for item in self do
+      return false if yield(item)
+    end
+    true
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```
- [ ] Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
 words.my_none?(/z/) == words.none?(/z/) #true
 words.my_none?(/d/) == words.none?(/d/) #true
```
- [ ] returns true if the block never returns true for all elements
```ruby
true_array = [nil, false, true, []]
false_array = [nil, false, nil, false]
false_array.my_none? #true
true_array.my_none? #false
```

File: my_enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    end
+    result
+  end
+
+  def my_any?
+    for item in self do
+      return true if yield(item)
+    end
+    false
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
true_array = [nil, false, true, []]
true_array.my_any? == true_array.any? #true
```
- [ ]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words = %w[dog door rod blade]
words.my_any?(Integer) == words.any?(Integer) #true
```
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true
```

File: my_enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
+module Enumerable
+  def my_each
+    for item in self do
+      yield(item)
+    end
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] my_each returns an Enumerator if no block is given
```ruby
array = Array.new(100) { rand(0...9) } }
array.my_each) == Enumerator # should return true
```

File: my_enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+  def my_each_with_index
+    i = 0
+    for item in self do
+      yield(item, i)
+      i += 1
+    end
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an enumerator if no block is given
```ruby
array.my_each_with_index == Enumerator # should return true
```

File: my_enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+  def my_select
+    new_arr = []
+    for item in self do
+      new_arr.push(item) if yield(item)
+    end
+    new_arr
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an enumerator if no block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_select == Enumerator # should return true
```

File: my_enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    true
+  end
+
+  def my_count
+    counter = 0
+    for item in self
+      counter += 1 if yield(item)
+    end
+    counter
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns the number of items in enum through enumeration
```ruby
array = Array.new(100){rand(0...9)}
array.my_count == array.count #true
```
- [ ] counts the number of items in enum that are equal to item if an argument is given
```ruby
array.my_count(0) == array.count(0) # should return true
```

File: my_enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+  def my_map
+    new_arr = []
+    for item in self do
+      new_arr << yield(item)
+    end
+    new_arr
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an Enumerator if no block is given
```ruby
array.my_map == Enumerator # should return true
```

File: my_enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+  def my_inject
+    sum = self[0]
+    for item in self do
+      sum = yield(sum, item)
+    end
+    sum
+  end
```
Comment (whiz25) [inline my_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) # should return true
```
- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
operation = proc { |sum, n| sum + n } 
search = proc { |memo, word| memo.length > word.length ? memo : word } 
array = Array.new(100){rand(0...9)}
words = %w[dog door rod blade]

array.my_inject(&operation) == array.inject(&operation) # should return true
actual = range.my_inject(4) { |prod, n| prod * n }
expected = range.inject(4) { |prod, n| prod * n }
actual == expected #true
words.my_inject(&search) == words.inject(&search) # should return true
```

File: my_enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      my_each do |item|
+        return false unless yield(item)
+      end
+    else
+      my_each do |item|
+        return false unless item
+      end
+    end
+    true
+  end
```
Comment (shshamim63) [inline my_enumerable.rb]: `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
[1, 2, 3].my_all?(Integer) #should return true
[1, 2, nil].my_all?(Integer) #should return false
```
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
%w[dog door rod blade].my_all?(/d/)   #should return true.
%w[dog door rod blade].my_all?(/t/)   #should return false.
```
- when a pattern other than Regex or a Class is given, returns true if all of the collection matches the pattern
```ruby
[3, 3, 3].my_all?(3) #should return true
[3, 3, 1].my_all?(3) #should return false
```

File: my_enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+      my_each do |item|
+        return true if yield(item)
+      end
+    elsif args.is_a? Hash
+      my_each do |x, y|
+        return true if yield(x, y)
+      end
+    end
+    false
+  end
```
Comment (shshamim63) [inline my_enumerable.rb]: `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns true if the block ever returns a value other than false or nil
```ruby
[1, 2, 3].my_any? { |num| num % 2 == 0 } #should return true
```
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
[1, 2, 3].my_any?(Integer) #should return true
[1, 'demo', false].my_any?(Integer) #should return true
['demo', false, nil].my_any?(Integer) #should return false
```
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_any?(/t/)    #should return false.
%w[dog door rod blade].my_any?(/d/)   #should return true.
```
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
[1, 2, 3].my_any?(3)   #should return true
[1, 2, 2].my_any?(3)   #should return false
```

File: my_enumerable.rb [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+      my_each do |item|
+        return false if yield(item)
+      end
+    elsif args.is_a? Hash
+      my_each do |y, z|
+        return false if yield(y, z)
+      end
+    end
+    true
+  end
```
Comment (shshamim63) [inline my_enumerable.rb]: `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns true if the block never returns true for all elements
```ruby
[1, 2, 3].my_none? { |num| num > 4 }  #should return true.
```
- when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2, 3].my_none?(String)  #should return true
['1', '2', '3'].my_none?(String)  #should return false
```
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/)   #should return true
%w[dog door rod blade].my_none?(/d/)   #should return false
```
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
[1, 2, 4].my_none?(3) #should return true
[1, 2, 3].my_none?(3) #should return false
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :tractor: 
___
Hi @jamezjaz :wave: 
Great job, you have covered the most challenging one my_inject, and I am sure that the issues I am gonna address in here gonna be too easy for you. Let's check the issues,
___
### :spades: Mandatory Changes :spades: 
___
1. Check the comment below and fix the edge cased related to the following methods.
- my_all?
- my_any?
- my_none?
___
#### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

File: my_enumerable.rb [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+  def my_all?(args = nil)
+    if args.nil?
+      my_each do |item|
+        return false unless item
+      end
+    elsif block_given?
+      my_each do |item|
+        return false unless yield(item)
+      end
+    else
```
Comment (Dandush03) [inline my_enumerable.rb]: As far as I can see the issue here is that you're missing the comparison needed to make it through...  
For example: 
return false unless item == 5. 
instead, you're just yielding the item. Hope this makes it clear now ;)

> Note: don't forget that in this method as my previous colleague mentioned it; it should return true if the block never returns false or nil 

###### Good Luck

**Review by Dandush03 — COMMENTED** [review]
# Status: :red_circle: Required Changes :red_circle: 
## Hey! :wave: 
## Great work so far. You're one step to finish it.

I can see that my colleague asked you for some correction that was not made.
I can understand the issue after reviewing the code and it's a very small thing to change. I will comment below to explain better the issue in one of the cases but it repeats the same issue for:

 - [ ] #my_all
 - [ ] #my_any
 - [ ] #my_none

Try to think about what I will explain below and if you need any help you can contact me on Monday through slack.

## Happy Coding :computer: :wink: 

### Daniel Laloush

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :tada: :tada: :tada: 
Great work @jamezjaz 
All required changes were implemented and now all methods are working as expected. :clap: 
Good luck with the next project

Happy coding :computer: 
**Sergio Torres**
